elastic/kibana
# [App Search] Analytics - add reusable Layout & Header + basic subroute view components

## Summary

Sorry for the size, I definitely did not realize how much code I was churning out until I opened the PR 💦 Definitely follow along by-commit if possible! I'll walk through QA + screenshots per-commit.

### AnalyticsUnavailable - ad300a5 

<img width="1181" alt="" src="https://user-images.githubusercontent.com/549407/104791090-487ab900-574e-11eb-8c03-3a01fe92ed6b.png">

#### QA:
- Setup: Either modify your ent-search rails server to send back a 400/500 response for the analytics endpoints, or update line 38 of `analytics_layout.tsx` to `if (true) return <AnalyticsUnavailable />;`
- Go to the following pages and confirm all of them show an analytics unavailable view:
- [x] http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/analytics
- [x] http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/analytics/top_queries
- [x] http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/analytics/query/test

### AnalyticsHeader - 7d08969

**Filters updating the URL search query:**

![pushing_search](https://user-images.githubusercontent.com/549407/104793484-a8299200-5757-11eb-8dd3-13295f69bd22.gif)

**Filters updated _from_ the URL search query:**

Tags:
![pulling_search_tag](https://user-images.githubusercontent.com/549407/104793482-a6f86500-5757-11eb-80c4-363bc3490258.gif)

Dates:
![pulling_search_dates](https://user-images.githubusercontent.com/549407/104793467-92b46800-5757-11eb-98f9-34a6e3f2b591.gif)

Responsive UI:
<img width="300" alt="" src="https://user-images.githubusercontent.com/549407/104791200-baeb9900-574e-11eb-84b3-938ce915ddbd.png">

### QA:
- Go to http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/analytics
- [x] Confirm that updating the tags and dates and then clicking the Apply Filters button correctly updates the URL query string
- [x] Refresh the page and confirm your selected tags & dates load back in from the URL bar
- [x] Confirm the log retention tooltip works correctly
- Open your devtools and track XHR calls.
- [x] Confirm that XHR calls to `/api/app_search/engines/{engine}/analytics` are being made on page load
- [x] Confirm that `?start=x&end=y&tag=z` search query params are being correctly forwarded to the API endpoint
- Go to [a query detail page](http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/analytics/query_detail/test)
- [x] Confirm that an XHR call to `/api/app_search/engines/{engine}/analytics/queries/{query}` is made
- [x] Confirm it correctly fwds search query params

### AnalyticsLayout & subroutes - 734e7fb & af13c62

<img width="1181" alt="" src="https://user-images.githubusercontent.com/549407/104793670-97c5e700-5758-11eb-9faa-a6e5c1708781.png">
<img width="1031" alt="" src="https://user-images.githubusercontent.com/549407/104793813-d0fe5700-5758-11eb-961c-d3e50fda4342.png">

#### QA

- Setup - turn off your Analytics logging in your settings
- Go to the following pages and confirm that they all show an analytics header and a log retention callout:
  - [x] http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/analytics
  - [x] http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/analytics/top_queries
  - [x] http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/analytics/top_queries_no_clicks
  - [x] http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/analytics/top_queries_with_clicks
  - [x] http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/analytics/top_queries_no_results
  - [x] http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/analytics/recent_queries

### QueryDetails view - f487596 & f035a46 

<img width="1155" alt="" src="https://user-images.githubusercontent.com/549407/104793863-10c53e80-5759-11eb-8939-cc0e17f6f33a.png">

#### QA:
- Go to http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/analytics/query_detail/test
- [x] Confirm that an analytics header and log retention callout renders as expected
- [x] Confirm the title of the page renders the text after `/query_detail/whatever`
- [x] Confirm that the breadcrumb trail renders as `Analytics > Query > whatever`
- [x] Confirm that clicking on the "Query" part of the breadcrumb trail does not 404, but takes you back to the `/analytics` overview

### Checklist
